other/web-js/client/src/data-api.js

```diff
@@ -0,0 +1,219 @@
+      return `https://${region}.${cloud}.data.mongodb-api.com/app/${appId}/endpoint/data/v1/`;
+    } else {
+      return `https://data.mongodb-api.com/app/${appId}/endpoint/data/v1/`;
+    }
+  }
+
+  // constructor(config: {
+  //   appId: string;
+  //   cloud?: string;
+  //   region?: string;
+  //   onAuthChange?: (currentUser: User) => void;
+  // }): DataApi;
+  constructor({ appId, cloud, region, onAuthChange }) {
+    this.appId = appId;
+    this.baseUrl = DataApi.constructBaseUrl(appId, { cloud, region });
+    this.client = new ClientApi({
+      appId,
+      cloud,
+      region,
+      onAuthChange: (newCurrentUser) => {
+        this.currentUser = newCurrentUser;
+        onAuthChange?.(this.currentUser);
+      },
+    });
+    this.currentUser = this.client.currentUser;
+  }
+
+  // registerUser(provider: string, credentials: any): Promise<void>
+  registerUser = async (provider, credentials) => {
+    await this.client.registerUser(provider, credentials);
+  };
+
+  // logIn(provider: string, credentials: any): Promise<void>
+  logIn = async (provider, credentials) => {
+    await this.client.logIn(provider, credentials);
+  };
+
+  // logOut(): Promise<void>
+  logOut = async () => {
+    await this.client.logOut();
+  };
+
+  /**
+   * An email/password authentication wrapper around the generic auth
+   * methods.
+   */
+  emailPasswordAuth = {
+    // registerUser(credentials: { email: string, password: string }): Promise<void>
+    registerUser: async ({ email, password }) => {
+      return await this.registerUser("local-userpass", { email, password });
+    },
+    // logIn(credentials: { email: string, password: string }): Promise<void>
+    logIn: async ({ email, password }) => {
+      return await this.logIn("local-userpass", { email, password });
+    },
+  };
+
+  /**
+   * Calls a requested Data API action endpoint.
+   * @param {string} action The name of a Data API action to perform, e.g. "insertOne"
+   * @param {object} input The request body for the action.
+   * @returns {object} The response body for the action.
+   */
+  async action(action, input) {
+    if (!this.currentUser) {
+      throw new Error(`Must be logged in to call a Data API action endpoint`);
+    }
+    // If the current user access token is expired, try to refresh the
+    // session and get a new access token.
+    await this.client.refreshExpiredAccessToken();
+
+    const url = new URL(`action/${action}`, this.baseUrl).href;
+    const resp = await fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/ejson",
+        Accept: "application/ejson",
+        Authorization: `Bearer ${this.currentUser.access_token}`,
+      },
+      body: JSON.stringify(input),
+    });
+    const response = await resp.json();
+
+    if (resp.status === 200 || resp.status === 201) {
+      return response;
+    } else {
+      throw new Error(response);
+    }
+  }
+
+  // type FindOneInput = {
```

Comment:
note that you can use JSDoc [@typedef](https://jsdoc.app/tags-typedef.html) here, which'd provide intellisense. though i think the TS types in comments are easier to read if you're actually looking at the source code